viral/two_photon.py

```diff
@@ -254,33 +256,236 @@ def place_cells(session: Cached2pSession, dff: np.ndarray) -> None:
-        session = Cached2pSession.model_validate_json(f.read())
-    print(f"Total number of trials: {len(session.trials)}")
-    print(
-        f"number of trials imaged {len([trial for trial in session.trials if trial_is_imaged(trial)])}"
+    # TODO: Think about step_size! Has to evenly divide the range, however we had it at 30 before (0-150; 150-180; ...)
+    first_position = 10
+    last_position = 170
+    step_size = 1
+    sampling_rate = 30
+    speed_positions = get_speed_positions(
+        position=position,
+        first_position=first_position,
+        last_position=last_position,
+        step_size=step_size,
+        sampling_rate=sampling_rate,
     )
 
-    dff = get_dff(mouse, date)
+    # Bin frames by speed
+    bin_size = 1
+    start = 0
+    max_speed = max(sp.speed for sp in speed_positions) + bin_size
+
+    dff_speed = list()
+    for bin_start in range(start, int(max_speed), bin_size):
+        if bin_start == 0:
+            frame_idx_bin = [
+                frame_position[idx] for idx, sp in enumerate(speed_positions) if sp.speed == 0
+            ]
+            # TODO: Is the handling of speed = 0 correct?
+            if len(frame_idx_bin) > 0:
+                dff_bin = dff[:, frame_idx_bin]
+                dff_speed.append(np.mean(dff_bin, axis=1))
+            else:
+                dff_speed.append(np.zeros(dff.shape[0]))  # handle no data for speed = 0
+            continue
+        frame_idx_bin = [
+            frame_position[idx]
+            for idx, sp in enumerate(speed_positions)
+            if bin_start <= sp.speed < bin_start + bin_size
+        ]
+        if len(frame_idx_bin) == 0:
+            dff_speed.append(np.zeros(dff.shape[0])) # speed = 0 gets included
+            continue
+        dff_bin = dff[:, frame_idx_bin]
+
+        if verbose:
+            print(f"bin_start: {bin_start}")
+            print(f"bin_end: {bin_start + bin_size}")
+            print(f"n_frames in bin: {len(frame_idx_bin)}")
+        dff_speed.append(np.mean(dff_bin, axis=1))
+
+    # print("\n")
+    return np.array(dff_speed).T
+
+
+def get_speed_activity(
+    trials: List[TrialInfo], dff: np.ndarray, wheel_circumference: float
+) -> np.ndarray:
+    test_matrices = list()
+
+    for _ in range(10):
+        train_idx = random.sample(range(len(trials)), len(trials) // 2)
+        test_idx = [idx for idx in range(len(trials)) if idx not in train_idx]
+
+        # Find the order in which to sort neurons in a random 50% of the trials
+        train_matrix = np.nanmean(
+            np.array(
+                [
+                    activity_trial_speed(trials[idx], dff, wheel_circumference)
+                    for idx in train_idx
+                ]
+            ),
+            axis=0,
+        )
+        train_matrix[:, 33 // 1 : 40 // 1] = 0
+        train_matrix[:, 76 // 1 : 85 // 1] = 0
+        train_matrix[:, 120 // 1 : 132 // 1] = 0
```

Comment:
Bizzare code from me lol, needs to be deleted.

Reply:
ooooh i think i'm trying to silence the landmark from the sorting, i'll go through this